### PR TITLE
Add support for navigation to page anchors

### PR DESCRIPTION
**Pull request by fauvel**

### CODE REVIEW

**Comment by fauvel** [comment]
@njam please review

**Comment by tomaszdurka** [comment]
Looks good, only a few things:

**Is it neccesary to have this anchor navigating feature in Layout/Abstract.js?**

**Fragment in request should be nullable.** You could compare fragment to null (not empty string) when generating url which is cleaner.
Only problem I see is with extracting fragment from request using `parse_url('...', PHP_URL_FRAGMENT)`:
  - `http://foo/bar#bar` fragment should be `'bar'`, is `'bar'`
  - `http://foo/bar` fragment should be `null`, is `null`
  - `http://foo/bar#` fragment should be `''`, is `null`

Introducing some additional trailing hash check could help. Wdyt?

**Comment by tomaszdurka** [comment]
When I think about it, maybe there should be nothing like empty string hash? I would just allow null for `setFragment()` then.

**Comment by njam** [comment]
Just talking with @vogdb who needs the same feature.
I don't think the hash should reach the server (it doesn't in HTTP). We could just keep it on the client side, and concatenate it with the URL coming back from the server before calling `replaceState()`?

**Comment by fauvel** [comment]
> I don't think the hash should reach the server (it doesn't in HTTP).

I agree that the hash doesn't **ought to** reach the web server. As a matter of fact (and as a best practice), most web browsers won't pass it along with the request when they handle the navigation by themselves (URL bar / HTML links). But – as a side note – there's nothing at the protocol level that doesn't allow you to do so; in fact, using the implementation I've suggested for our ajax navigation, you actually get the fragment in `$_SERVER['REQUEST_URI']` (at least in my dev box, with nginx and our php-fpm config).

So I also agree we should remove support for fragments in request URLs on the server side.

To handle it on the client side, we could (a) remove the fragment from the request URL before loading the page in the ajax navigation, (b) check whether the URL of the page response matches the request URL (without fragment), and (c) in this case – and in this case only! – re-add the fragment to the URL of the page response before storing it in the navigation history, and jump to the anchor.

Is this what you have in mind?

**Comment by vogdb** [comment]
He definitely meant that. Lets not wait any minute longer and finish this issue! =)

**Comment by fauvel** [comment]
@njam @tomaszdurka @vogdb please review!

File: library/CM/App.js [hunk 1]
```diff
@@ -42,8 +42,8 @@ var CM_App = CM_Class_Abstract.extend({
       return view;
     }
     return _.find(this.views, function(view) {
-      return view.hasClass(className);
-    }) || null;
+        return view.hasClass(className);
```
Comment (vogdb) [inline library/CM/App.js]: indent typo?

Comment (fauvel) [inline library/CM/App.js]: indent fix :)

**Comment by vogdb** [comment]
- Should we use `responsePath` instead of `pathResponse` if we use the names as `modulePath`, `templatePath` and etc. in other places?
- Lets be more js-style!
```js
if ('' !== window.location.hash) {
```
with
```js
if (window.location.hash) {
```

**Comment by fauvel** [comment]
@njam @tomaszdurka good to merge?

**Comment by vogdb** [comment]
lgtm!

**Comment by njam** [comment]
- CM/Layout/Abstract.js: Now that we check if the server changed the fragment, we can call `replaceState()` only if necessary?
 - Maybe introduce a boolean variable to clarify on this - `responsePathChange`?
- CM/Layout/Abstract.js: Move the anchor-scrolling into `_onPageSetup()`?
- CM/Layout/Abstract.js: You renamed "fragment" to "path". While I agree it's probably the better choice I think we should keep it consistent. At the moment anything after the domain and port we call "fragment". I'd suggest to stick with it?

**Comment by fauvel** [comment]
> Now that we check if the server changed the fragment, we can call `replaceState()` only if necessary?

Are you sure that the path passed to `loadPage` is in the navigation history already?

**Comment by vogdb** [comment]
Ok. If we decided to be perfect then lets do a replacement:
```js
      var anchor = document.getElementById(hash);
      if (!anchor) {
        var anchorList = document.getElementsByName(hash);
        if (anchorList.length) {
          anchor = anchorList[0];
        }
      }
```
with
```js
var anchor = $('#' + hash)[0] || $('[name=' + hash + ']')[0];
```
or
```js
var anchor = $('#' + hash).get(0) || $('[name=' + hash + ']').get(0);
```
whatever you prefer =)

**Comment by njam** [comment]
> Are you sure that the path passed to loadPage is in the navigation history already?

I think so, but maybe it's safer and more robust to keep it!

**Comment by fauvel** [comment]
@vogdb I always think about code injection when I see things like that: `'[name=' + hash + ']'` – I would rather keep the dom implementation. @njam good to merge?

**Comment by njam** [comment]
agree - http://youmightnotneedjquery.com/ :D

Should we test this on some *edge case devices* first? - https://github.com/cargomedia/sk/wiki/Supported-Browsers

@vogdb maybe you can help with that?

**Comment by vogdb** [comment]
>  I always think about code injection when I see things like that:
> agree - http://youmightnotneedjquery.com/ :D

So, why to include jQuery in CM codebase? if we have a library then we should use it on 100% otherwise don't use it at all.
In this case I don't see how the injection can be done properly. The processing of hash happens only on the client side.

**Comment by vogdb** [comment]
> @vogdb maybe you can help with that?

Will test on Desktop browsers + my readl Android + iPhone simulators.

**Comment by njam** [comment]
Agree there's no real injection possible. It's just a sizzle selector, and we only use it to get the DOM element!

**Comment by fauvel** [comment]
I don't see a big security threat here either, and I have absolutely nothing against jQuery! In the worst case, someone could forge a link to our website with a hash that would break the jQuery selector, that wouldn't be the end of the world. But it just hurts the eye to see that we're not escaping user input. And escaping it would make the whole construct so complicated, that I would prefer to stay with the straight-forward, fuzzy-free DOM approach. You really don't need the whole jQuery overhead to handle such a simple thing as the selection of an anchor. There's also quite a few jQuery implementation of the scrolling to the anchor itself, but `scrollIntoView()` is really cross-browser compatible afaik. @vogdb if it works in all browsers we're supporting, I would advocate to keep it as is. But I agree it's only a matter of taste.

**Comment by vogdb** [comment]
Ok. Then lets not prolongate the issue! Lgtm!

**Comment by vogdb** [comment]
Oh. Wait. Need to test it carefully.

**Comment by fauvel** [comment]
Yes, I'd rather be more productive and discuss less on such details, too, but maybe we'll learn something from your testing?

**Comment by vogdb** [comment]
You should test on the edge Android devices that we have in the office.

**Comment by vogdb** [comment]
Ok. Works fine. Unfortunately I don't have a device with Android Browser =( My Android device is too new and already has replaced it with Mobile Chrome.

**Comment by fauvel** [comment]
So good to merge? @njam agree?

**Comment by njam** [comment]
So we should still test it on Android?
@NicolasSchmutz could you do that?

**Comment by NicolasSchmutz** [comment]
It took a while but it looks good on all Android devices we have.

**Comment by fauvel** [comment]
Tagged 1.104.1
